pkg/manager/infosync/info.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2023 PingCAP, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package infosync
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+	"path"
+	"time"
+
+	"github.com/pingcap/TiProxy/lib/config"
+	"github.com/pingcap/TiProxy/lib/util/retry"
+	"github.com/pingcap/TiProxy/lib/util/sys"
+	"github.com/pingcap/TiProxy/lib/util/waitgroup"
+	"github.com/pingcap/TiProxy/pkg/util/versioninfo"
+	"github.com/pingcap/errors"
+	"github.com/siddontang/go/hack"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/client/v3/concurrency"
+	"go.uber.org/zap"
+)
+
+const (
+	tiproxyTopologyPath = "/topology/tiproxy"
+
+	topologySessionTTL    = 45
+	topologyRefreshIntvl  = 30 * time.Second
+	topologyPutTimeout    = 2 * time.Second
+	topologyPutRetryIntvl = 1 * time.Second
+	topologyPutRetryCnt   = 3
+	logInterval           = 10
+
+	TTLSuffix  = "ttl"
+	InfoSuffix = "info"
+)
+
+// InfoSyncer syncs TiProxy topology to ETCD.
+// It writes 2 items: `/topology/tiproxy/.../info` and `/topology/tiproxy/.../ttl`.
+// `info` is written once and `ttl` will be erased automatically after TiProxy is down.
+// The code is modified from github.com/pingcap/tidb/domain/infosync/info.go.
+type InfoSyncer struct {
+	syncConfig      syncConfig
+	lg              *zap.Logger
+	etcdCli         *clientv3.Client
+	wg              waitgroup.WaitGroup
+	cancelFunc      context.CancelFunc
+	topologySession *concurrency.Session
+}
+
+type syncConfig struct {
+	sessionTTL    int
+	refreshIntvl  time.Duration
+	putTimeout    time.Duration
+	putRetryIntvl time.Duration
+	putRetryCnt   uint64
+}
+
+type TopologyInfo struct {
+	Version        string `json:"version"`
+	GitHash        string `json:"git_hash"`
+	IP             string `json:"ip"`
+	Port           string `json:"port"`
+	StatusPort     string `json:"status_port"`
+	DeployPath     string `json:"deploy_path"`
+	StartTimestamp int64  `json:"start_timestamp"`
+}
+
+func NewInfoSyncer(etcdCli *clientv3.Client, lg *zap.Logger) *InfoSyncer {
+	return &InfoSyncer{
+		etcdCli: etcdCli,
+		lg:      lg,
+		syncConfig: syncConfig{
+			sessionTTL:    topologySessionTTL,
+			refreshIntvl:  topologyRefreshIntvl,
+			putTimeout:    topologyPutTimeout,
+			putRetryIntvl: topologyPutRetryIntvl,
+			putRetryCnt:   topologyPutRetryCnt,
+		},
+	}
+}
+
+func (is *InfoSyncer) Init(ctx context.Context, cfg *config.Config) error {
+	topologyInfo, err := is.getTopologyInfo(cfg)
+	if err != nil {
+		is.lg.Error("get topology failed", zap.Error(err))
+		return err
+	}
+
+	childCtx, cancelFunc := context.WithCancel(ctx)
+	is.cancelFunc = cancelFunc
+	is.wg.Run(func() {
+		is.updateTopologyLivenessLoop(childCtx, topologyInfo)
+	})
+	return nil
+}
+
+func (is *InfoSyncer) updateTopologyLivenessLoop(ctx context.Context, topologyInfo *TopologyInfo) {
+	// We allow TiProxy to start before PD, so do not fail in the main goroutine.
+	if err := is.initTopologySession(ctx); err != nil {
+		return
+	}
+	_ = is.storeTopologyInfo(ctx, topologyInfo)
```

Comment:
So we just ignore errors?

Reply:
Yes, the log is printed. We can just retry in the next loop. Failing to store topology is not a critical problem currently.
Do you think we should try infinitely?

Reply:
Can we make then noreturn, for example:

```
func store(..) {
err := retry.RetryNotify(..)
log(err)
}
```

Reply:
Now I wrapped them into `syncTopology` and put the log outside of `storeTopologyInfo` and `updateTopologyAliveness`.